hoomd/hpmc/pytest/test_clusters.py

```diff
@@ -0,0 +1,233 @@
+                                    seed=1)
+
+    setattr(cl, attr, value)
+    assert getattr(cl, attr) == value
+
+
+@pytest.mark.parametrize("attr,value", valid_attrs)
+def test_valid_setattr_attached(attr, value, simulation_factory,
+                                two_particle_snapshot_factory,
+                                valid_args):
+    """Test that Clusters can get and set attributes while attached."""
+
+    integrator = valid_args[0]
+    args = valid_args[1]
+    # Need to unpack union integrators
+    if isinstance(integrator, tuple):
+        inner_integrator = integrator[0]
+        integrator = integrator[1]
+        inner_mc = inner_integrator(23456)
+        for i in range(len(args["shapes"])):
+            # This will fill in default values for the inner shape objects
+            inner_mc.shape["A"] = args["shapes"][i]
+            args["shapes"][i] = inner_mc.shape["A"]
+    mc = integrator(23456)
+    mc.shape["A"] = args
+    mc.shape["B"] = args
+
+    cl = hoomd.hpmc.update.Clusters(trigger=hoomd.trigger.Periodic(10),
+                                    swap_type_pair=['A', 'B'], seed=1)
+    dim = 2 if 'polygon' in integrator.__name__.lower() else 3
+    sim = simulation_factory(two_particle_snapshot_factory(particle_types=['A', 'B'],
+                                                           dimensions=dim, d=2, L=50))
+    sim.operations.updaters.append(cl)
+    sim.operations.integrator = mc
+
+    sim.run(0)
+
+    setattr(cl, attr, value)
+    assert getattr(cl, attr) == value
+
+
+@pytest.mark.parametrize("delta_mu", [-2.0, 0, 2.0])
+def test_swap_moves(delta_mu, simulation_factory,
+                    lattice_snapshot_factory):
+    """Test that Clusters can swap types."""
+
+    sim = simulation_factory(lattice_snapshot_factory(particle_types=['A', 'B'],
+                                                      dimensions=3, a=4, n=7, r=0.1))
+
+
+    mc = hoomd.hpmc.integrate.Sphere(seed=1, d=0.1, a=0.1)
+    mc.shape['A'] = dict(diameter=1)
+    mc.shape['B'] = dict(diameter=1)
+    sim.operations.integrator = mc
+
+    cl = hoomd.hpmc.update.Clusters(trigger=hoomd.trigger.Periodic(5),
+                                    swap_type_pair=['A', 'B'],
+                                    swap_move_ratio=1,
+                                    delta_mu=delta_mu,
+                                    seed=12)
+    sim.operations.updaters.append(cl)
+
+    # set every other particle to type B (typeid=1)
+    snap = sim.state.snapshot
+    if snap.exists:
+        snap.particles.typeid[range(0, sim.state.N_particles, 2)] = 1
+    sim.state.snapshot = snap
+
+    # number of particles of each type should change after a run
+    sim.run(100)
+
+    snap = sim.state.snapshot
+    if snap.exists:
+        num_type_B = np.sum(snap.particles.typeid)
+        num_type_A = len(snap.particles.typeid) - num_type_B
+        acceptance = cl.swap_moves[0]/np.sum(cl.swap_moves)
```

Comment:
Accessing properties like `swap_moves` must be done outside the `if snap.exists`. These are collective operations that must be performed on all ranks. What is happening here is that a message from `swap_moves` on one rank is getting received by `broadcast_box` on the other rank which skipped past this conditional.

Reply:
Whoops, I totally missed that. Seems to be working now. Thanks!